# Принятый в команде Code Style

--- 

## Импорты классов

Все полные наименования классов должны быть заменены на использование оператора `use` в начале файла.

### Пример

Вместо:
```php
$response = new \Parser\Task\Domain\Response\TaskResponse();
```

Используйте:
```php
use Parser\Task\Domain\Response\TaskResponse;

// ...

$response = new TaskResponse();
```

### Исключения

- Не добавляйте оператор `use` для встроенных типов и функций PHP (int, string, count(), etc.)
- Не добавляйте оператор `use` для классов в том же namespace

--- 

## Рекомендации по оформлению кода

1. **При написании тестов:**
    - Использовать статические вызовы методов PHPUnit, например (
          self::assertTrue вместо $this->assertTrue,
          self::assertFalse вместо $this->assertFalse,
          self::assertSame вместо $this->assertSame,
          self::assertNull вместо $this->assertNull,
          self::assertEquals вместо $this->assertEquals,
      )

2. **При написании кода:**
    - Следить за длиной строк (не более 120 символов)
    - Разбивать длинные строки на несколько строк для улучшения читаемости
    - Не переопределять методы без изменения логики
    - Всегда проверять форматирование кода после внесения изменений
    - Использовать типизированные константы для улучшения type safety

3. **Документирование кода** (PHPDoc):
    - Добавить PHPDoc блоки для всех public методов
    - Документировать параметры (@param) только для описания массивов
    - Документировать возвращаемое значение (@return) только для описания массивов
    - Документировать возможные исключения (@throws)
    - Добавлять типы массивов в PHPDoc в строковых комментариях для улучшения анализа кода
    - Для скалярных массивов в PHPDocs вместо `array<...>` указывать `list<...>`
    - Комментарии в коде должны быть написаны на русском языке

   **Исключения**:
    - Комментарии могут содержать английские термины, относящиеся к коду (например, @param, @return, @throws)
    - Названия классов, методов и переменных в комментариях должны сохранять оригинальное написание

    Пример:
    ```php
    /**
    * Трансформирует DTO из API в доменную Entity.
    *
    * @param list<TicketApiDto> $dto DTO из Freshdesk API
    *
    * @return array<string, Ticket> Трансформированная сущность
    *
    * @throws InvalidTicketData Если данные не соответствуют требованиям
    */
    public function transformFromApi(array $dto, string $message): array
    {
        /** @var array<string, Ticket> $result */
        $result = [];
        // .... 
    }
    ```

4. **При работе с Entity и DTO:**
    - Использовать публичные readonly свойства вместо геттеров (согласно возможностям PHP 8.5)

5. **При работе с интерфейсами:**
    - Удалять лишние (@return void) из PHPDoc комментариев
    - Следить за корректностью описания методов
